pkg/cmd/provisioning/identity_provider.go

```diff
@@ -0,0 +1,335 @@
+package provisioning
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	awssdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	jose "gopkg.in/square/go-jose.v2"
+
+	"github.com/openshift/cloud-credential-operator/pkg/aws"
+)
+
+var (
+	// IdentityProviderOpts captures the options that affect creation of the identity provider
+	IdentityProviderOpts = options{
+		InfraName:     "",
+		Region:        "",
+		PublicKeyPath: "",
+	}
+
+	// discoveryDocumentTemplate is a template of the discovery document that needs to be populated with appropriate values
+	discoveryDocumentTemplate = `{
+	"issuer": "%s",
+	"jwks_uri": "%s/%s",
+    "response_types_supported": [
+        "id_token"
+    ],
+    "subject_types_supported": [
+        "public"
+    ],
+    "id_token_signing_alg_values_supported": [
+        "RS256"
+    ],
+    "claims_supported": [
+        "aud",
+        "exp",
+        "sub",
+        "iat",
+        "iss",
+        "sub"
+    ]
+}`
+)
+
+type JSONWebKeySet struct {
+	Keys []jose.JSONWebKey `json:"keys"`
+}
+
+func createIdentityProvider(client aws.Client, infraName, region, publicKeyPath, targetDir string) error {
+	bucketName := fmt.Sprintf("%s-installer", infraName)
+	issuerURL := fmt.Sprintf("https://%s.s3.%s.amazonaws.com", bucketName, region)
+
+	var bucketTagValue string
+	_, err := client.CreateBucket(&s3.CreateBucketInput{
+		Bucket: awssdk.String(bucketName),
+	})
+	if err != nil {
+		var aerr awserr.Error
+		if errors.As(err, &aerr) {
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyOwnedByYou:
+				bucketTagValue = sharedCcoctlAWSResourceTagValue
```

Comment:
Help me understand these proposed tag values. If we create the bucket we call it "owned", but if the tagging fails and we we-run the tool, we then get tagged "shared"?
I don't think this is a huge deal, but what are we getting from these tagging values?

Reply:
@abhinavdahiya do you know the details on this?

Reply:
IIUC, we should be tagging "shared" if the object already exists (so that any uninstall process does not try to delete it), and "owned" if ccoctl created the object.
But how do we handle the installation of cluster1 where ccoctl creates everything, then perhaps the install for cluster2 is run with the same infra/name prefix, ccoctl should do what?
a) unmark the "owned" and set it to "shared" when it sees the pre-existing objects?
b) something else?

Reply:
I don't see why we want to distinguish owned vs shared?
shared would be useful when multiple different OIDC providers were using the same s3 bucket, but it looks like there is one bucket per OIDC provider, so there doesn't seem to be requirement for shared.

I would say just keep it owned for now..

Reply:
On a related note, in #304 I'm adding the ability to just output JSON instead of making the AWS API calls (that you can then `aws iam identity-provider --cli-input-json` with). Since tagging is a separate operation, should we generate JSON files, or assume since the user is manually handling/reviewing the creation of the objects that they will set up whatever tagging they want?